magma/lib/magma/loader.rb

```diff
@@ -69,15 +69,172 @@ def to_payload
       return payload.to_hash
     end
 
+    def is_collection_attribute?(attribute)
+      attribute.is_a?(Magma::CollectionAttribute)
+    end
+
+    def is_child_attribute?(attribute)
+      attribute.is_a?(Magma::ChildAttribute)
+    end
+
+    def is_table_attribute?(attribute)
+      attribute.is_a?(Magma::TableAttribute)
+    end
+
+    def is_incoming_link?(attribute)
+      # Based on the attribute class, determines if this is the incoming attribute,
+      #   for a foreign-key relationship. If so, return true.
+      is_collection_attribute?(attribute) ||
+      is_child_attribute?(attribute) ||
+      is_table_attribute?(attribute)
+    end
+
+    def explicit_child_revision_exists?(revisions, model_name, record_name, attribute_name)
+      # Check if the child's parent / link is being explicitly set by the user
+      !!revisions.dig(model_name.to_sym, record_name.to_sym, attribute_name.to_sym)
+    end
+
+    def explicit_link_revision_exists?(revisions, model_name, record_name, attribute, child_record_name)
+      # We need to check if the child record has
+      #   been set for ANY record in the same parent_model.
+
+      # NOTE: different check behavior for CollectionAttribute vs. ChildAttribute!
+      explicit_revisions_found = 0
+
+      revisions[model_name.to_sym].each do |record_name, record_revisions|
+        next unless record_revisions.key?(attribute.attribute_name.to_sym)
+
+        attribute_revision = record_revisions[attribute.attribute_name.to_sym]
+
+        explicit_revisions_found += 1 if is_collection_attribute?(attribute) && attribute_revision.include?(child_record_name)
+        explicit_revisions_found += 1 if is_child_attribute?(attribute) && child_record_name == attribute_revision
+      end
+
+      explicit_revisions_found > 0
+    end
+
+    def explicit_revision_exists?(revisions:, parent_model:, parent_record_name:, child_record_name:, parent_attribute:)
+      # Cannot find explicit revisions from @records, because
+      #   some of those are calculated! So we have to look for
+      #   explicit revisions from the user-supplied revisions hash.
+      # Note that because explicit revisions can be either parent / link -> child
+      #   or child -> parent / link, we need to check both cases.
+      explicit_child_revision_exists?(
+        revisions,
+        parent_attribute.link_model.model_name,
+        child_record_name,
+        parent_model.model_name) ||
+      explicit_link_revision_exists?(
+        revisions,
+        parent_model.model_name,
+        parent_record_name,
+        parent_attribute,
+        child_record_name)
+    end
+
+    def push_implicit_link_revisions(revisions)
+      # When updating link or parent attributes from the top-down,
```

Comment:
The main benefit of doing this here seems to be to collect inquiries across the update, otherwise this might even fit inside `revision_to_links`, if it could be aware of the existing records; but since we don't want to query for those one-at-a-time, this method becomes necessary.

Another possibility I considered was, the loader first actually loads the current data for all items in the revision. This data can then be made available as required to revision_to_links, which would have no trouble computing these implicit links in place. You could also calculate which parts of the revision were redundant (that is, not an update) much more easily, and probably thereby avoid setting `updated_at` unnecessarily.

I think what you have written is the most efficient route, but I thought the idea of always preloading is interesting - updates shouldn't be extremely common, so the overhead might be worth it.

Reply:
Would pre-loading be less efficient if the validation turns out to reject the updates?